### PR TITLE
Add Array implementation

### LINKED ISSUE #3: Implement Array

### PR DESCRIPTION
**Pull request by hellovietduc**
Closes: #3

### CODE REVIEW

File: array/array.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+func (a *Array) ValueAt(index int) (int, error) {
+	if index < 0 || index >= a.Size() {
+		return -1, errors.New("Index is out of range")
+	}
+	return a.arr[index], nil
+}
+
+func (a *Array) Insert(item int, index int) error {
+	if index < 0 || index >= a.Capacity() {
```
Comment (trungducc) [inline array/array.go]: Valid range is from `0` to `a.size` so `index >= a.Capacity()` must be `index >= a.size`

File: array/array.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	for i := a.size; i > index; i-- {
+		a.arr[i] = a.arr[i-1]
+	}
+
+	a.arr[index] = item
+
+	// insert at index out of size range
+	// will increase the size to index + 1
+	if index >= a.Size() {
+		a.size = index + 1
```
Comment (trungducc) [inline array/array.go]: This case will never happen.

File: array/array.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+func (a *Array) Pop() (int, error) {
+	if a.IsEmpty() {
+		return -1, errors.New("Array is empty")
+	}
+
+	a.checkAndReduce()
+
+	a.size--
+	value := a.arr[a.size]
```
Comment (trungducc) [inline array/array.go]: At this point, `a.arr[a.size]` is an invalid value. I prefer to getting `value` before decreasing `size`.

File: array/array.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+	return value, nil
+}
+
+func (a *Array) RemoveAt(index int) (int, error) {
+	if index < 0 || index >= a.Capacity() {
```
Comment (trungducc) [inline array/array.go]: Same as above, `a.size` should be used instead of `a.Capacity()`

File: array/array.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
+package array
+
+import (
+	"errors"
+)
+
+type Array struct {
+	arr  []int
```
Comment (trungducc) [inline array/array.go]: I think this property should not be named `arr`. We don't keep an array inside another array. `data` can be better name.

File: array/array.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+func (a *Array) checkAndShrink() {
+	capacity := len(a.arr)
+	if a.size < capacity {
+		return
+	}
+
+	// size = capacity, double the capacity
+	newArr := make([]int, capacity*2)
```
Comment (trungducc) [inline array/array.go]: We don't keep magic numbers in our code. Grow and shrink factor should be defined as constants so it can be easier to update in the future.

File: array/array.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+import (
+	"errors"
+)
+
+type Array struct {
+	arr  []int
+	size int
+}
+
+func NewArray(capacity int) *Array {
```
Comment (trungducc) [inline array/array.go]: We should have another constructor or give `capacity` a default value. 8 or 16 will be a good default capacity.

File: array/array.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+func (a *Array) Size() int {
+	return a.size
+}
+
+func (a *Array) Capacity() int {
+	return len(a.arr)
+}
+
+func (a *Array) IsEmpty() bool {
+	return a.Size() == 0
```
Comment (trungducc) [inline array/array.go]: Should we use `a.size` instead of `a.Size()`?

File: array/array.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	if a.IsEmpty() {
+		return -1, errors.New("Array is empty")
+	}
+
+	a.checkAndReduce()
+
+	a.size--
+	value := a.arr[index]
```
Comment (trungducc) [inline array/array.go]: I prefer to getting `value` before decreasing `size`.

**Review by trungducc — CHANGES_REQUESTED** [review]
Which coding convention are you using? Please help to check tab size should be 6 or not. Also need to add comment for describing functions's purpose

File: array/array.go (hunk 1 above)
Comment (hellovietduc) [reply array/array.go]: The intention of this function is to allow inserting at any index in range 0 to capacity - 1, which mimics the behavior in JavaScript.

For example, in JS I can do:
```javascript
const arr = []
arr[6] = 1
console.log(arr.length) // output: 7
```
Every other elements from index 0 to 5 are now `undefined`.

So in my implementation, if I do:
```go
arr := NewArray(10)
arr.Insert(1, 6)
fmt.Print(arr.Size()) // output: 7
```
Every other elements from index 0 to 5 would be `0`, since Go `int` type can't be `nil`.

File: array/array.go (hunk 2 above)
Comment (hellovietduc) [reply array/array.go]: Same as my comment above, this is to jump the size of the array when inserting an element **outside** of size range, but **inside** the capacity range.

File: array/array.go (hunk 3 above)
Comment (hellovietduc) [reply array/array.go]: I will not fail, because I decrease the size before accessing.
Before I used to write like this:
```go
value := a.arr[a.size - 1]
a.size--
```
I just think the language has to do 2 subtract operations here (`a.size - 1` and `a.size--`), so I do the subtraction first then access the value later.

File: array/array.go (hunk 4 above)
Comment (hellovietduc) [reply array/array.go]: Same as above too! If my implementation allows inserting outside of size range - inside of capacity range, then deleting operation should behave the same.

File: array/array.go (hunk 5 above)
Comment (hellovietduc) [reply array/array.go]: Naming things is always difficult @@ I'll edit this.

File: array/array.go (hunk 8 above)
Comment (hellovietduc) [reply array/array.go]: What if I change something in the logic of computing the array size? I think it's better to change it in one function only.

**Comment by hellovietduc** [comment]
> Which coding convention are you using? Please help to check tab size should be 6 or not. Also need to add comment for describing functions's purpose

The default Go extension in VSCode does the lint style for me. I'm not sure if it uses spaces or tabs, seems like tabs.

Anyways, thanks for all your thorough reviews 🤩

File: array/array.go [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+		arr: make([]int, capacity),
+	}
+}
+
+func (a *Array) Size() int {
+	return a.size
+}
+
+func (a *Array) Capacity() int {
+	return len(a.arr)
```
Comment (thanks173) [inline array/array.go]: Maybe we should use the builtin function `cap`

File: array/array.go (hunk 1 above)
Comment (trungducc) [reply array/array.go]: Actually JS arrays don't have a concept of size but we have. We implement array to understand how an array works in most languages and JS is an edge case. Inserting at an index which is greater than size is an unexpected behavior. We don't know which value will be automatically inserted. Normally, calling `insert` once will only add a value into array but in this case, `size` can be increased from 0 to 7 after an insertion.

@thanks173 How do you think?

File: array/array.go (hunk 3 above)
Comment (trungducc) [reply array/array.go]: It will not fail but values which are not in range [0, size] are invalid because we don't manage them. If you still want to decrease size before getting the value, you can `return a.arr[--a.size]` if it's available or at least don't create `value` variable.

File: array/array.go (hunk 5 above)
Comment (trungducc) [reply array/array.go]: Another way is naming `arr` property as `items` so each value in array is an item.

File: array/array.go (hunk 8 above)
Comment (trungducc) [reply array/array.go]: As I understand, `Size()` is a getter for `size`. If you change something in the logic of computing the array size, `size` should be changed instead of computing again in getter

Comment (hellovietduc) [reply array/array.go]: It actually is a getter in this case, but I still think there's no problem in using a getter inside the struct implementation. Just like how I use `IsEmpty()`.

File: array/array.go (hunk 3 above)
Comment (hellovietduc) [reply array/array.go]: Well, `--a.size` in Go throws an error :(

For the `value` part, as I said, I try to minic JS implementation so when you delete an item at an index, you get the value in that index as returned value.

File: array/array.go (hunk 8 above)
Comment (trungducc) [reply array/array.go]: Nothing is wrong if you use `Size()` but it will take 1 more step to get the value. It's up to you.

File: array/array.go (hunk 3 above)
Comment (trungducc) [reply array/array.go]: Can we do it without creating `value` variable? How about `return arr[size], nil`?

File: array/array.go (hunk 1 above)
Comment (hellovietduc) [reply array/array.go]: I understand the purpose of learning to understand how array works. My implementation looks like a wrapper class with enhance functionality, I guess. :)

File: array/array.go (hunk 3 above)
Comment (hellovietduc) [reply array/array.go]: But the item will be lost, because I shift the array to the left right below that.

Comment (trungducc) [reply array/array.go]: I think you are misunderstanding. This is `pop` function and you don't need to shift anything here. There is only one more line in this function after this line `return value, nil`. Maybe you can check again.

File: array/array.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+func (a *Array) IsEmpty() bool {
+	return a.Size() == 0
+}
+
+func (a *Array) ValueAt(index int) (int, error) {
+	if index < 0 || index >= a.Size() {
```
Comment (trungducc) [inline array/array.go]: In this function, range is from 0 to size but in another function range is from 0 to capacity. It will make user confuse when using.

File: array/array.go (hunk 8 above)
Comment (hellovietduc) [reply array/array.go]: Oh I see! My bad. Didn't notice this.

File: array/array.go (hunk 3 above)
Comment (hellovietduc) [reply array/array.go]: I must have been copy-pasted from the Remove function here. :( I see the mistake.

File: array/array.go (hunk 1 above)
Comment (thanks173) [reply array/array.go]: In my point of view, JS is a mess. We should not use it as a best practice.

File: array/array.go (hunk 10 above)
Comment (hellovietduc) [reply array/array.go]: I don't understand the differences between `cap` and `len` when calling them on an array yet. Any notes?

File: array/array.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+)
+
+// Array is a struct for storing linear equal-size data.
+type Array struct {
+	items []int
+	size  int
+}
+
+// NewArray creates a new Array instance.
+func NewArray(args ...int) *Array {
```
Comment (trungducc) [inline array/array.go]: Does this function accept more than 1 parameter?

File: array/array.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+// Capacity returns the current maximum number of items
+// that can be stored in the Array.
+func (a *Array) Capacity() int {
+	return len(a.items)
+}
+
+// IsEmpty returns whether the Array has no items or not.
+func (a *Array) IsEmpty() bool {
```
Comment (trungducc) [inline array/array.go:43]: I think you don't need to mention function name at the beginning of comment. For example: `Returns true if the array is not empty. Otherwise, return false`

File: array/array.go [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+	if index < 0 || index >= a.size {
+		return EmptyValue, errors.New(indexOutOfRange)
+	}
+	return a.items[index], nil
+}
+
+// Insert adds the value to the given index and shift
+// the items from that index to the right.
+func (a *Array) Insert(value int, index int) error {
+	if index < 0 || index >= a.size {
```
Comment (trungducc) [inline array/array.go]: It should be `index < 0 || index > a.size`. It's ok to insert an item at index 8 while array is having 8 items.

File: array/array.go [hunk 15]
```diff
@@ -0,0 +1,144 @@
+package array
+
+import (
+	"errors"
+)
+
+const (
+	// EmptyValue is a constant indicates that
+	// a value returned from the Array is empty.
+	EmptyValue       = 0
```
Comment (trungducc) [inline array/array.go:10]: Default value should be a value which is rarely used. I suggest to use min value of `int` type instead of 0.

File: array/array_test.go [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+	})
+
+	t.Run("Append more values so array must grow in capacity", func(t *testing.T) {
+		t.Logf("Capacity before: %d", arr.Capacity())
+
+		for i := 4; i < 20; i++ {
+			arr.Append(i)
+		}
+
+		t.Logf("Capacity after: %d", arr.Capacity())
```
Comment (trungducc) [inline array/array_test.go]: We will not log anything in test cases cause noone will read it. Instead of logging capacity, it's better to check capacity is equal a value. If it's not equal expected value, test case fails.

**Review by trungducc — CHANGES_REQUESTED** [review]
I am not sure about the way test cases are written in Go. @thanks173 could you please take a look?

File: array/array.go (hunk 12 above)
Comment (hellovietduc) [reply array/array.go]: This is optional parameters in Go. If nothing is passed, then args will be `nil`.

File: array/array.go (hunk 15 above)
Comment (hellovietduc) [reply array/array.go:10]: I remember @thanks173 gave a review saying it can be 0, because 0 is a falsey value.

File: array/array.go (hunk 13 above)
Comment (hellovietduc) [reply array/array.go:43]: image
I don't know but it throws a warning like this. Seems like a convention for Go too.
What's your opinion @thanks173 ?

Comment (thanks173) [reply array/array.go:43]: It's a convention.

File: array/array.go (hunk 12 above)
Comment (trungducc) [reply array/array.go]: I mean what will happen if there is more than 1 parameter is passed. If you only need 1 parameter, why do we need using `...`?

Comment (hellovietduc) [reply array/array.go]: It's the only way to create a function which receives optional parameters. Without the `...`, I always have to pass something.

Comment (trungducc) [reply array/array.go]: Why do you **have to** use optional parameters in this case?

Comment (hellovietduc) [reply array/array.go]: Because somebody may want to create an Array instance with different initial capacity than the default one, right?

Comment (trungducc) [reply array/array.go]: Why don't you just have 2 constructors, 1 with capacity parameter and 1 without capacity parameter? 
If you use optional parameters and user calls `NewArray(0, 1, 2, 3, 4, 5, 6)`, what will happen? Is it unexpected behavior?

Comment (hellovietduc) [reply array/array.go]: Then the second constructor must have a different name than `NewArray`. That's kinda messy. I think I can check carefully what user passes in.

Comment (hellovietduc) [reply array/array.go]: Or, just don't let users pass their expected initial capacity at all?

File: array/array.go [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// Pop removes that last item in the Array.
+func (a *Array) Pop() int {
+	if a.IsEmpty() {
+		return EmptyValue
+	}
+
+	if a.getLoadFactor() < lowerLoadFactor {
+		a.shrink()
```
Comment (trungducc) [inline array/array.go:114]: Should we shirk in case array is empty? I think we should have min capacity.

File: .github/workflows/go.yml [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+    branches:
+      - '*'
+      - '**'
+      - '!master'
+
+jobs:
+  run_tests:
+    name: Run tests
+    runs-on: ubuntu-latest
+    steps:
```
Comment (trungducc) [inline .github/workflows/go.yml]: I think this part should be in another PR.

**Review by trungducc — CHANGES_REQUESTED** [review]

File: .github/workflows/go.yml (hunk 18 above)
Comment (hellovietduc) [reply .github/workflows/go.yml]: This file triggers GitHub Actions on my fork repo.

File: array/array.go (hunk 17 above)
Comment (hellovietduc) [reply array/array.go:114]: I have a check empty before this.

File: .github/workflows/go.yml (hunk 18 above)
Comment (trungducc) [reply .github/workflows/go.yml]: If you want to add this file, create another PR. This PR doesn't relate to GitHub Action.

**Review by trungducc — APPROVED** [review]

**Review by thanks173 — APPROVED** [review]
